src/client/dfs/dfs_sys.c

```diff
@@ -1399,6 +1399,49 @@ dfs_sys_mkdir(dfs_sys_t *dfs_sys, const char *dir, mode_t mode,
 	return rc;
 }
 
+int
+dfs_sys_mkdir_p(dfs_sys_t *dfs_sys, const char *dir_path, mode_t mode, daos_oclass_id_t cid)
+{
+	int         path_len = strnlen(dir_path, PATH_MAX);
+	char       *_path    = NULL;
+	char       *ptr     = NULL;
+	int         rc       = 0;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+	if (dir_path == NULL)
+		return EINVAL;
+	if (path_len == PATH_MAX)
+		return ENAMETOOLONG;
+
+	D_STRNDUP(_path, dir_path, path_len);
+	if (_path == NULL)
+		return ENOMEM;
+
+	/* iterate through the parent directories and create them if necessary */
+	for (ptr = _path + 1; *ptr != '\0'; ptr++) {
+		if (*ptr != '/')
+			continue;
+
+		/* truncate the string here to create the parent */
+		*ptr = '\0';
+		rc   = dfs_sys_mkdir(dfs_sys, _path, mode, cid);
```

Comment:
i think there is another inefficiency here with this loop. every iteration is passing the full path right?
for example if dir_path is /a/b/c/d/e
we are calling dfs_sys_mkdir with
/a
/a/b
/a/b/c
...

This is also doing unnecessary path walks every iteration. (please correct me if i understood your parsing wrong).
If this is the case, then it would be much more efficient to not use dfs_sys_mkdir() here and use dfs_mkdir() instead or even dfs_open with O_CREAT and DIR type to create the directories and use the parent from the previous iteration to avoid the intermediate lookups in every iteration.

Reply:
I was thinking about this... But perhaps I misunderstood the point of the hash table in dfs_sys. @daltonbohning: Shouldn't the parent objects be pulled out of that instead of having to do the lookups?

Reply:
The dfs sys hash table is to avoid looking up the same path again. But `hash_lookup` doesn't recursively cache. So this works okay
```
dfs_sys_mkdir /a     # lookup a, cache a
dfs_sys_mkdir /a/b   # a from cache, lookup b, cache b
dfs_sys_mkdir /a/b/c # b from cache, lookup c, cache c
```
but the reverse is less optimal
```
dfs_sys_mkdir /a/b/c  # none cached, dfs_lookup traverses a,b,c
                      # cache c - ONLY c is cached!
dfs_sys_mkdir /a/b    # none cached! dfs_lookup traverses a,b
                      # cache b
```

Reply:
But if caching is turned off, Mohamad's point is correct.
I guess it's possible to optimize for both cases, but that's more work and probably out of scope :)

Reply:
I added some debugging lines and it seems like it's doing what we expect:
```
[ RUN      ] DFS_SYS_UNIT_TEST14: DFS Sys mkdir_p
daos ERR  src/client/dfs/dfs_sys.c:204 hash_lookup() hash hit: /
daos ERR  src/client/dfs/dfs_sys.c:208 hash_lookup() hash miss: /a
daos ERR  src/client/dfs/dfs_sys.c:204 hash_lookup() hash hit: /
daos ERR  src/client/dfs/dfs_sys.c:204 hash_lookup() hash hit: /a/b
daos ERR  src/client/dfs/dfs_sys.c:204 hash_lookup() hash hit: /
daos ERR  src/client/dfs/dfs_sys.c:204 hash_lookup() hash hit: /a
daos ERR  src/client/dfs/dfs_sys.c:204 hash_lookup() hash hit: /a/b
```

Given that, does it alleviate your performance concern, @mchaarawi?

Reply:
As Dalton indicated, caching is optional. so you will still see unnecessary lookups when you have nested dirs here. but we can improve that later, so i won't hold this up bec of that.
The path parsing and existence check + type check can be combined together if we implement something similar within dfs itself. I will make a note to add something along that lines when i get a chance.